### PR TITLE
Check index params when create index

### PR DESCRIPTION
**Pull request by xiaocai2333**
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
/kind improvement

issue: #21539

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proxy/index_params.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	case DefaultStringIndexType, DefaultIndexType:
+		return checkIndexParamsUtil(ScalarIndexParams, indexParams)
+	default:
+		return fmt.Errorf("IndexType is not support: %s", indexType)
+	}
+}
+
+func checkIndexParamsUtil(expected map[string]struct{}, actual map[string]string) error {
+	for key := range actual {
+		if _, ok := expected[key]; !ok {
```
Comment (czs007) [inline internal/proxy/index_params.go:80]: So the logic of the check is to report an error if indexParams contains an unexpected key.
If indexParams does not include the expected key, is it not checked?

Comment (xiaocai2333) [reply internal/proxy/index_params.go:80]: We have additional logic `checkTrain` to check this.

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/proxy/task_index.go [hunk 2]
```diff
@@ -227,36 +234,31 @@ func fillDimension(field *schemapb.FieldSchema, indexParams map[string]string) e
...
 		log.Warn("Field data type don't support the index build type", zap.String("fieldDataType", field.GetDataType().String()), zap.String("indexType", indexType))
 		return fmt.Errorf("field data type %s don't support the index build type %s", field.GetDataType().String(), indexType)
 	}
 
-	ok = adapter.CheckTrain(indexParams)
-	if !ok {
-		log.Warn("Create index with invalid params", zap.Any("index_params", indexParams))
-		return fmt.Errorf("invalid index params: %v", indexParams)
+	if err = adapter.CheckTrain(indexParams); err != nil {
+		log.Warn("Create index with invalid params", zap.Any("index_params", indexParams), zap.Error(err))
```
Comment (longjiquan) [inline internal/proxy/task_index.go:255]: Just use `Info` here. Try not to use `zap.Any`.

Comment (longjiquan) [reply internal/proxy/task_index.go:255]: Do not capitalize the first letter.

File: internal/proxy/task_index.go [hunk 3]
```diff
@@ -227,36 +234,31 @@ func fillDimension(field *schemapb.FieldSchema, indexParams map[string]string) e
...
+		return err
+	}
+
+	if err = adapter.CheckUnnecessaryParams(indexParams); err != nil {
+		log.Warn("Create index with invalid params", zap.Error(err))
```
Comment (longjiquan) [inline internal/proxy/task_index.go:260]: Just use `Info` here. Do not capitalize the first letter.

File: internal/util/indexparamcheck/index_params.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		OutgoingEdgeSize: {}, IncomingEdgeSize: {}}
+	RHNSWPQParams = map[string]struct{}{common.IndexTypeKey: {}, common.MetricTypeKey: {}, common.DimKey: {},
+		HNSWM: {}, EFConstruction: {}, PQM: {}}
+	ScalarIndexParams = map[string]struct{}{common.IndexTypeKey: {}}
+)
+
+func checkUnnecessaryParams(expected map[string]struct{}, actual map[string]string) error {
+	for key := range actual {
+		if _, ok := expected[key]; !ok {
+			return fmt.Errorf("invalid index param: %s", key)
```
Comment (longjiquan) [inline internal/util/indexparamcheck/index_params.go:160]: Use `merr` to wrap the leaf error.

File: internal/util/indexparamcheck/utils.go [hunk 5]
```diff
@@ -47,11 +48,16 @@ func CheckIntByRange(params map[string]string, key string, min, max int) bool {
 //   1. the key does not exist, or
 //   2. the data does not appear in the container
 // Return true otherwise
-func CheckStrByValues(params map[string]string, key string, container []string) bool {
+func CheckStrByValues(params map[string]string, key string, container []string) error {
 	value, ok := params[key]
 	if !ok {
-		return false
+		return fmt.Errorf("lack necessary parameter: %s", key)
```
Comment (longjiquan) [inline internal/util/indexparamcheck/utils.go:54]: Use `merr` to wrap leaf error.

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Comment by longjiquan** [comment]
Remember to cherr-pick this feature to master after this merged.

**Pull request closed without merging**